File: src/main/java/main/ui/renderer/ChipListCellRenderer.java
package main.ui.renderer;

import main.App;
import main.puzzle.Chip;
import main.ui.resource.AppImage;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;

/**
 *
 * @author Bunnyspa
 */
public class ChipListCellRenderer extends DefaultListCellRenderer {

    private final App app;
    private final boolean factored;
    private final double factor;

    public ChipListCellRenderer(App app) {
        super();
        this.app = app;
        this.factored = false;
        this.factor = 1.0;
    }

    public ChipListCellRenderer(App app, double factor) {
        super();
        this.app = app;
        factored = true;
        this.factor = factor;
    }

    @NotNull
    @Override
    public Component getListCellRendererComponent(JList<?> list, @Nullable Object value, int index, boolean isSelected, boolean cellHasFocus) {
        DefaultListCellRenderer cr = (DefaultListCellRenderer) super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus);

        cr.setHorizontalAlignment(CENTER);
        cr.setText("");

        Chip c = (Chip) value;
        if (value != null) {
            ImageIcon icon = AppImage.Chip.get(app, c);
            if (!factored) {
                cr.setIcon(icon);
            } else {
                Image image = icon.getImage().getScaledInstance((int) (icon.getIconWidth() * factor), (int) (icon.getIconHeight() * factor), Image.SCALE_SMOOTH);
                cr.setIcon(new ImageIcon(image));
            }
            cr.setBackground(isSelected ? Color.LIGHT_GRAY : Color.WHITE);
        } else {
            cr.setIcon(null);
            cr.setBackground(Color.WHITE);
        }
        return cr;
    }
}

File: src/main/java/main/ui/renderer/ChipListCellRenderer.kt
package main.ui.renderer

import main.App
import main.puzzle.Chip
import main.ui.resource.AppImage
import java.awt.Color
import java.awt.Component
import java.awt.Image
import javax.swing.DefaultListCellRenderer
import javax.swing.ImageIcon
import javax.swing.JList

/**
 *
 * @author Bunnyspa
 */
class ChipListCellRenderer : DefaultListCellRenderer {
    private val app: App
    private val factored: Boolean
    private val factor: Double

    constructor(app: App) : super() {
        this.app = app
        factored = false
        factor = 1.0
    }

    constructor(app: App, factor: Double) : super() {
        this.app = app
        factored = true
        this.factor = factor
    }

    override fun getListCellRendererComponent(
        list: JList<*>?,
        value: Any?,
        index: Int,
        isSelected: Boolean,
        cellHasFocus: Boolean
    ): Component {
        val cr =
            super.getListCellRendererComponent(list, value, index, isSelected, cellHasFocus) as DefaultListCellRenderer
        cr.horizontalAlignment = CENTER
        cr.text = ""
        val c = value as Chip?
        if (value != null) {
            val icon = AppImage.Chip[app, c!!]
            if (!factored) {
                cr.icon = icon
            } else {
                val image = icon.image.getScaledInstance(
                    (icon.iconWidth * factor).toInt(),
                    (icon.iconHeight * factor).toInt(),
                    Image.SCALE_SMOOTH
                )
                cr.icon = ImageIcon(image)
            }
            cr.background = if (isSelected) Color.LIGHT_GRAY else Color.WHITE
        } else {
            cr.icon = null
            cr.background = Color.WHITE
        }
        return cr
    }
}